### PR TITLE
Add SSM way to connect linux workspace

### LINKED ISSUE #29: [Feature Request] securely access Linux instance in regions ec2-connect not available
**Is your feature request related to a problem? Please describe.**

SWB is using [EC2 instance connect](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/Connect-using-EC2-Instance-Connect.html) to securely access Linux instances provisioned in workspace. However ec2 instance connect feature is not available in all regions.

**Describe the solution you'd like**

Provide another secure solution to SSH access the Linux instances in workspace.

### PR DESCRIPTION
**Pull request by dengmingtong**
Issue #, if available:

closes #29 

Description of changes:

Checklist:

- [yes ] Have you successfully deployed to an AWS account with your changes?
- [ yes] Have you successfully deployed to an AWS China account with your changes?
- [ yes] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully tested with your changes locally?
- [ ] If new dependencies have been added, have they been pinned to specific versions?
- [ ] Is this change also required on the AWS Solution version?
- [ ] Have you updated openapi.yaml if you made updates to API definition (including add, delete or update parameter and request data schema)?
- [ ] If you had to run manual tests, have you considered automating those tests by adding them to [end-to-end tests](../main/end-to-end-tests/README.md)?
- [ ] If you are updating the changelog and vending out a new release, have you updated versionNumber and versionDate in [.defaults.yml](../main/config/settings/.defaults.yml)

AS review ticket id:

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: addons/addon-base-raas-ui/packages/base-raas-ui/src/parts/environments-sc/parts/ScEnvironmentSsmConnections.js [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    const isSsm = scheme => scheme === 'ssm';
+    const connections = this.environment.getConnections(item => isSsm(item.scheme));
+    return connections;
+  }
+
+  handleConnect = id =>
+    action(async () => {
+      const store = this.getConnectionStore();
+      const connections = this.environment.connections;
+      const connectInfo = _.find(connections, ['id', id]) || {};
```
Comment (zxkane) [inline addons/addon-base-raas-ui/packages/base-raas-ui/src/parts/environments-sc/parts/ScEnvironmentSsmConnections.js:59]: prevent opening more than one SSM connect window? does it work?

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-connection-service.js [hunk 2]
```diff
@@ -220,6 +229,90 @@ class EnvironmentScConnectionService extends Service {
...
+
+  async generateSigninSSMConsoleLink(accessKeyId, secretAccessKey, sessionToken, instanceId, inputSessionDuration) {
+    const session = {
+      sessionId: accessKeyId,
+      sessionKey: secretAccessKey,
+      sessionToken,
+    };
+
+    let sessionDuration = 3599; // default session duration, 59:59
+    if (Number(inputSessionDuration) > 0) {
```
Comment (zxkane) [inline addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-connection-service.js]: also should check the maximum allowed value of session duration

**Review by zxkane — CHANGES_REQUESTED** [review]

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/__tests__/environment-sc-connection-service.test.js [hunk 3]
```diff
@@ -503,6 +503,48 @@ describe('EnvironmentScConnectionService', () => {
...
+        return {
+          accessKeyId: 'accessKeyId',
+          secretAccessKey: 'secretAccessKey',
+          sessionToken: 'sessionToken',
+        };
+      });
+
+      service.generateSigninSSMConsoleLink = jest.fn(
+        () =>
+          'https://us-east-1.signin.aws.amazon.com/federation?Action=login&Issuer=Instruqt&Destination=https://us-east-1.console.aws.amazon.com/systems-manager/session-manager/i-xxx?region=us-east-1&SigninToken=xxx',
```
Comment (zxkane) [inline addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/__tests__/environment-sc-connection-service.test.js]: Mocking the to be tested method with values does not make sense.
You should mock the aws sdk call then testing the other logics in that method.

See https://aws.amazon.com/blogs/developer/mocking-modular-aws-sdk-for-javascript-v3-in-unit-tests/ for mocking aws sdk api.

**Review by zxkane — CHANGES_REQUESTED** [review]

File: addons/addon-base-raas-ui/packages/base-raas-ui/src/parts/environments-sc/parts/ScEnvironmentSsmConnections.js (hunk 1 above)
Comment (dengmingtong) [reply addons/addon-base-raas-ui/packages/base-raas-ui/src/parts/environments-sc/parts/ScEnvironmentSsmConnections.js:59]: Currently, we cannot prevent opening more than one window,  implementation is the same with Sagemaker.
Allow user to open multi windows

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-connection-service.js (hunk 2 above)
Comment (dengmingtong) [reply addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-connection-service.js]: I already limit maximum in linux Cloudformation template input parameter。

Comment (zxkane) [reply addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-connection-service.js]: The workspace template could be developed by third party, adding a sanity checking still makes sense.

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-connection-service.js [hunk 4]
```diff
@@ -220,6 +229,90 @@ class EnvironmentScConnectionService extends Service {
...
+      consoleDomainSuffix = 'aws.amazon.com';
+    }
+
+    const getSigninTokenURL = `https://${region}.signin.${consoleDomainSuffix}/federation`;
+
+    const response = await axios.get(getSigninTokenURL, { params });
+
+    const signinToken = response.data.SigninToken;
+
+    const ssmConsoleUrl = `https://${region}.signin.${consoleDomainSuffix}/federation?Action=login&Issuer=Instruqt&Destination=https://${region}.console.${consoleDomainSuffix}/systems-manager/session-manager/${instanceId}?region=${region}&SigninToken=${signinToken}`;
```
Comment (zxkane) [inline addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-connection-service.js]: the fixed issuer `Instruqt` is required? or any meaning?

File: main/config/settings/.defaults.yml [hunk 5]
```diff
@@ -257,3 +257,8 @@ defaultIdpType: cognito_user_pool
 enableAmiSharing: false
 devopsProfile: devops
 # ======================================================================
+
+# ==================== Region List which do not supporting EC2 Instance Connect feature ====================
+# ==================== Refer: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-connect-methods.html ===============
+noEc2ConnectRegions: '["cn-northwest-1", "cn-north-1", "ap-east-1", "ap-northeast-3", "af-south-1", "eu-south-1", "me-south-1"]'
```
Comment (zxkane) [inline main/config/settings/.defaults.yml:263]: how did you get those list? from documentation?

Comment (dengmingtong) [reply main/config/settings/.defaults.yml:263]: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-connect-methods.html
Supported in all AWS Regions except .....

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-connection-service.js (hunk 4 above)
Comment (dengmingtong) [reply addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-connection-service.js]: Yes, Issuer means the organization that issues the URL, It is not mandatory. 
And Instruqt is not appositly.
I removed it.

File: addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-connection-service.js (hunk 2 above)
Comment (dengmingtong) [reply addons/addon-base-raas/packages/base-raas-services/lib/environment/service-catalog/environment-sc-connection-service.js]: Sure, added it.

**Review by chenhaiyun — APPROVED** [review]

**Review by zxkane — APPROVED** [review]

**Pull request merged**